### PR TITLE
TASK: Thumbnail generator fixes

### PR DESCRIPTION
**Pull request by aertmann**
- reuse existing thumbnails
- return original asset if suitable as thumbnail
- do no silence errors during publishing of thumbnails
- remove unnecessary SVG generator
- styling of asset editor

NEOS-1165

### CODE REVIEW

**Comment by aertmann** [comment]
:+1:

**Comment by aertmann** [comment]
hmm I see #286 introduced usage of it again.. that change shouldn't have been merged, big :-1:.. anyway will take care of it

**Comment by aertmann** [comment]
:+1:

File: TYPO3.Neos/Configuration/Settings.yaml [hunk 1]
```diff
@@ -423,9 +423,6 @@ TYPO3:
       'TYPO3.Neos:Preview':
         maximumWidth: 1000
         maximumHeight: 1000
-      'TYPO3.Neos:IconPreview':
-        maximumWidth: 16
-        maximumHeight: 16
```
Comment (kdambekalns) [inline TYPO3.Neos/Configuration/Settings.yaml]: So why did you remove this again? Anything that speaks against this?

Comment (kdambekalns) [inline TYPO3.Neos/Configuration/Settings.yaml]: Ok, you removed the usage above. Hm, then I don't get the usecase for the presets. I thought they would pregenerate thumbnails, and 16x16 is needed for the AssetList in the inspector…

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php [hunk 2]
```diff
@@ -194,11 +207,6 @@ public function getUriForThumbnail(Thumbnail $thumbnail)
             ), 1450178437);
         }
 
-        try {
-            list($package, $path) = $this->resourceManager->getPackageAndPathByPublicPath($staticResource);
-            return $this->resourceManager->getPublicPackageResourceUri($package, $path);
-        } catch (Exception $exception) {
-            return $staticResource;
-        }
+        return $this->resourceManager->getPublicPackageResourceUriByPath($staticResource);
```
Comment (kdambekalns) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php]: Good one, such an easy fix.

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php [hunk 3]
```diff
@@ -86,6 +86,19 @@ class ThumbnailService
...
+        // Calculates the dimensions of the thumbnail to be generated and returns the thumbnail image if the new
+        // dimensions differ from the specified image dimensions, otherwise the original image is returned.
+        if ($asset instanceof ImageInterface) {
+            if ($asset->getWidth() === null && $asset->getHeight() === null) {
+                return $asset;
+            }
+            $maximumWidth = ($configuration->getMaximumWidth() > $asset->getWidth()) ? $asset->getWidth() : $configuration->getMaximumWidth();
+            $maximumHeight = ($configuration->getMaximumHeight() > $asset->getHeight()) ? $asset->getHeight() : $configuration->getMaximumHeight();
+            if ($configuration->isUpScalingAllowed() === false && $maximumWidth === $asset->getWidth() && $maximumHeight === $asset->getHeight()) {
+                return $asset;
```
Comment (kdambekalns) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php]: Here and in the return statement above an Asset is returned, but the return value is documented to be a thumbnail. That has been mentioned by @dfeyer already and should be fixed.

File: TYPO3.Neos/Configuration/Settings.yaml (hunk 1 above)
Comment (aertmann) [inline TYPO3.Neos/Configuration/Settings.yaml]: yes, IMO it's better to reuse the existing thumbnails (250x250) than generating separate icons for all of them.. there's a much higher chance those thumbnails exists already, ensuring that calls like this one and for link/reference searches don't waste time on generating thumbnails.. so the speed gain of smaller images is outweighed by fewer needed thumbnails, larger re-use and high probability of those thumbnails already existing. I didn't add it in the first place for the other places where small icons are used as well (media list view & link search)

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php (hunk 3 above)
Comment (aertmann) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php]: changed it to return image interface since this allows skipping unnecessary thumbnails and removes the flaky SVG thumbnail generator, which would easily break since it copies the path to the resource and stores separately. Additionally that creates a requirement of URI building meaning the base uri needs to be determined, breaking CLI compatibility.

**Comment by aertmann** [comment]
added styling of the asset editor

**Comment by aertmann** [comment]
:+1:

**Comment by kdambekalns** [comment]
Ok, thanks for the follow-up. Agreed, and :+1: by reading. Can only test later today/tomorrow…

File: TYPO3.Neos/Configuration/Settings.yaml (hunk 1 above)
Comment (kitsunet) [inline TYPO3.Neos/Configuration/Settings.yaml]: WE might think about this again for mobile because there the file size might matter, but I would be fine to use the bigger thumbs for now.

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbnailGenerator/FontDocumentThumbnailGenerator.php [hunk 4]
```diff
@@ -75,7 +75,7 @@ public function refresh(Thumbnail $thumbnail)
             $red = imagecolorallocate($im, 0xFF, 0xFF, 0xFF);
             $black = imagecolorallocate($im, 0x00, 0x00, 0x00);
 
-            imagefilledrectangle($im, 0, 0, 1000, 1000, $red);
+            imagefilledrectangle($im, 0, 0, $width, $height, $red);
```
Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbnailGenerator/FontDocumentThumbnailGenerator.php]: With the fixed positions for the preview texts this will probably not work out?

Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbnailGenerator/FontDocumentThumbnailGenerator.php]: But the resize below should take care to bring it down to the right size I think?

**Comment by kitsunet** [comment]
basically like, just left one comment about the font preview where I think that might not work out... But will test it

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbnailGenerator/FontDocumentThumbnailGenerator.php (hunk 4 above)
Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbnailGenerator/FontDocumentThumbnailGenerator.php]: Huh, seems to work fine, so ok :)

**Comment by kitsunet** [comment]
Checked, tested, works :+1: merging

**Comment by kitsunet** [comment]
Thanks for taking care :)
